crypto/dh/dh_lib.c

```diff
@@ -246,14 +246,21 @@ void DH_get0_pqg(const DH *dh, BIGNUM **p, BIGNUM **q, BIGNUM **g)
 int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 {
     /* q is optional */
-    if (p == NULL || g == NULL)
+    if ((p == NULL && dh->p == NULL)
+        || (g == NULL && dh->g == NULL))
```

Comment:
Considering what's being said on openssl-dev, I wonder if that test (and the other corresponding ones in this change) shouldn't be like this instead:

``` C
    if ((p == NULL && dh->p == NULL) || (p != NULL && dh->p != NULL)
        || (g == NULL && dh->g == NULL) || (g != NULL && dh->g != NULL))
```

That is, either the DH field isn't initialised and the corresponding parameter *must* be non-NULL, or the DH field *is* initialised and the corresponding parameter *must* be NULL.